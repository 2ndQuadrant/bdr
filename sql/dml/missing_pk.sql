-- test sanity checks for tables without pk
SELECT * FROM public.bdr_regress_variables()
\gset

\c :writedb1

SELECT bdr.bdr_replicate_ddl_command($$
	CREATE TABLE public.bdr_missing_pk_parent(a serial PRIMARY KEY);
	CREATE TABLE public.bdr_missing_pk(a serial) INHERITS (public.bdr_missing_pk_parent);
	CREATE VIEW public.bdr_missing_pk_view AS SELECT * FROM public.bdr_missing_pk;
$$);

INSERT INTO bdr_missing_pk SELECT generate_series(1, 10);
SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);
\c :readdb2
SELECT * FROM bdr_missing_pk;

-- these should fail
\c :writedb2
UPDATE bdr_missing_pk SET a = 1;
DELETE FROM bdr_missing_pk WHERE a = 1;

UPDATE bdr_missing_pk_parent SET a = 1;
DELETE FROM bdr_missing_pk_parent WHERE a = 1;

WITH foo AS (
	UPDATE bdr_missing_pk SET a = 1 WHERE a > 1 RETURNING a
) SELECT * FROM foo;

WITH foo AS (
	DELETE FROM bdr_missing_pk RETURNING a
) SELECT * FROM foo;

UPDATE bdr_missing_pk_view SET a = 1;
DELETE FROM bdr_missing_pk_view WHERE a = 1;

WITH foo AS (
	UPDATE bdr_missing_pk_view SET a = 1 WHERE a > 1 RETURNING a
) SELECT * FROM foo;

WITH foo AS (
	DELETE FROM bdr_missing_pk_view RETURNING a
) SELECT * FROM foo;

WITH foo AS (
	UPDATE bdr_missing_pk SET a = 1 RETURNING *
) INSERT INTO bdr_missing_pk SELECT * FROM foo;

WITH foo AS (
	DELETE FROM bdr_missing_pk_view RETURNING a
) INSERT INTO bdr_missing_pk SELECT * FROM foo;


-- success again
TRUNCATE bdr_missing_pk;
SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);
\c :readdb1
SELECT * FROM bdr_missing_pk;

\c :writedb1
SELECT bdr.bdr_replicate_ddl_command($$DROP TABLE public.bdr_missing_pk CASCADE;$$);
