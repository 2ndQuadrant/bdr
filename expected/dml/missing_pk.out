-- test sanity checks for tables without pk
SELECT * FROM public.bdr_regress_variables()
\gset
\c :writedb1
BEGIN;
SET LOCAL bdr.permit_ddl_locking = true;
SELECT bdr.bdr_replicate_ddl_command($$
	CREATE TABLE public.bdr_missing_pk_parent(a serial PRIMARY KEY);
	CREATE TABLE public.bdr_missing_pk(a serial) INHERITS (public.bdr_missing_pk_parent);
	CREATE VIEW public.bdr_missing_pk_view AS SELECT * FROM public.bdr_missing_pk;
$$);
NOTICE:  merging column "a" with inherited definition
CONTEXT:  during DDL replay of ddl statement: 
	CREATE TABLE public.bdr_missing_pk_parent(a serial PRIMARY KEY);
	CREATE TABLE public.bdr_missing_pk(a serial) INHERITS (public.bdr_missing_pk_parent);
	CREATE VIEW public.bdr_missing_pk_view AS SELECT * FROM public.bdr_missing_pk;

 bdr_replicate_ddl_command 
---------------------------
 
(1 row)

COMMIT;
INSERT INTO bdr_missing_pk SELECT generate_series(1, 10);
SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);
 pg_xlog_wait_remote_apply 
---------------------------
 
(1 row)

\c :readdb2
SELECT * FROM bdr_missing_pk;
 a  
----
  1
  2
  3
  4
  5
  6
  7
  8
  9
 10
(10 rows)

-- these should fail
\c :writedb2
UPDATE bdr_missing_pk SET a = 1;
ERROR:  Cannot run UPDATE or DELETE on table bdr_missing_pk because it does not have primary key.
HINT:  Add primary key to the table
DELETE FROM bdr_missing_pk WHERE a = 1;
ERROR:  Cannot run UPDATE or DELETE on table bdr_missing_pk because it does not have primary key.
HINT:  Add primary key to the table
UPDATE bdr_missing_pk_parent SET a = 1;
ERROR:  Cannot run UPDATE or DELETE on table bdr_missing_pk because it does not have primary key.
HINT:  Add primary key to the table
DELETE FROM bdr_missing_pk_parent WHERE a = 1;
ERROR:  Cannot run UPDATE or DELETE on table bdr_missing_pk because it does not have primary key.
HINT:  Add primary key to the table
WITH foo AS (
	UPDATE bdr_missing_pk SET a = 1 WHERE a > 1 RETURNING a
) SELECT * FROM foo;
ERROR:  Cannot run UPDATE or DELETE on table bdr_missing_pk because it does not have primary key.
HINT:  Add primary key to the table
WITH foo AS (
	DELETE FROM bdr_missing_pk RETURNING a
) SELECT * FROM foo;
ERROR:  Cannot run UPDATE or DELETE on table bdr_missing_pk because it does not have primary key.
HINT:  Add primary key to the table
UPDATE bdr_missing_pk_view SET a = 1;
ERROR:  Cannot run UPDATE or DELETE on table bdr_missing_pk because it does not have primary key.
HINT:  Add primary key to the table
DELETE FROM bdr_missing_pk_view WHERE a = 1;
ERROR:  Cannot run UPDATE or DELETE on table bdr_missing_pk because it does not have primary key.
HINT:  Add primary key to the table
WITH foo AS (
	UPDATE bdr_missing_pk_view SET a = 1 WHERE a > 1 RETURNING a
) SELECT * FROM foo;
ERROR:  Cannot run UPDATE or DELETE on table bdr_missing_pk because it does not have primary key.
HINT:  Add primary key to the table
WITH foo AS (
	DELETE FROM bdr_missing_pk_view RETURNING a
) SELECT * FROM foo;
ERROR:  Cannot run UPDATE or DELETE on table bdr_missing_pk because it does not have primary key.
HINT:  Add primary key to the table
WITH foo AS (
	UPDATE bdr_missing_pk SET a = 1 RETURNING *
) INSERT INTO bdr_missing_pk SELECT * FROM foo;
ERROR:  Cannot run UPDATE or DELETE on table bdr_missing_pk because it does not have primary key.
HINT:  Add primary key to the table
WITH foo AS (
	DELETE FROM bdr_missing_pk_view RETURNING a
) INSERT INTO bdr_missing_pk SELECT * FROM foo;
ERROR:  Cannot run UPDATE or DELETE on table bdr_missing_pk because it does not have primary key.
HINT:  Add primary key to the table
-- success again
TRUNCATE bdr_missing_pk;
SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);
 pg_xlog_wait_remote_apply 
---------------------------
 
(1 row)

\c :readdb1
SELECT * FROM bdr_missing_pk;
 a 
---
(0 rows)

\c :writedb1
BEGIN;
SET LOCAL bdr.permit_ddl_locking = true;
SELECT bdr.bdr_replicate_ddl_command($$DROP TABLE public.bdr_missing_pk CASCADE;$$);
NOTICE:  drop cascades to view public.bdr_missing_pk_view
CONTEXT:  during DDL replay of ddl statement: DROP TABLE public.bdr_missing_pk CASCADE;
 bdr_replicate_ddl_command 
---------------------------
 
(1 row)

COMMIT;
